### PR TITLE
Add thenCallback method

### PR DESCRIPTION
**Pull request by APshenkin**
thenCallback method allow to complete some operations with request before sending response.

example of usage:
```
await server.get("/mocked-endpoint").thenCallback(req => {
 if (req.body.myVar = 'foo') {
  return {status: 200, body: {obj: req.body.myVar}, headers: {myHeader: 'foo'}}
 }
 return {status: 404, body: 'not found'}
});
```

Wait for your thought and suggestions

this PR should be merged after https://github.com/pimterry/mockttp/pull/1

### CODE REVIEW

File: src/rules/handlers.ts [hunk 1]
```diff
@@ -48,24 +59,76 @@ export function buildHandler
...
+                hostname: request.hostname,
+                path: request.path,
+                headers: request.headers,
+                body: { buffer, text, json, formData }
+            }
+            let ourResponse;
+            try {
+                ourResponse = await callback(cleanRequest);
+            } catch (err) {
+                throw err;
```
[the remark this answers is not in the document]
Comment (APshenkin) [reply src/rules/handlers.ts]: What failing status code will be best in this case?

File: src/rules/handlers.ts [hunk 2]
```diff
@@ -48,24 +59,76 @@ export function buildHandler
...
+            }
+            const cleanRequest = {
+                protocol: request.protocol,
+                method: request.method,
+                url: request.url,
+                hostname: request.hostname,
+                path: request.path,
+                headers: request.headers,
+                body: { buffer, text, json, formData }
+            }
```
[the remark this answers is not in the document]
Comment (APshenkin) [reply src/rules/handlers.ts]: this is a bad name for variables :) Will rename it

File: src/rules/handlers.ts [hunk 3]
```diff
@@ -48,24 +59,76 @@ export function buildHandler
...
+            let ourResponse;
+            try {
+                ourResponse = await callback(cleanRequest);
+            } catch (err) {
+                throw err;
+            }
+            if (typeof ourResponse.body === 'object') {
+                ourResponse.body = JSON.stringify(ourResponse.body);
```
[the remark this answers is not in the document]
Comment (APshenkin) [reply src/rules/handlers.ts]: I think, that this should be out of this PR. We need to think how to complete this better for different types of response

**Comment by APshenkin** [comment]
Just one more tip about why this feature is necessary.

If you want to mock any SOAP service, than you have to implement logic in callback to receive different responses in different cases. `thenReplyWithTemplate ` will not allow to do that. Because of this I created this PR. In our project we want to mock SOAP service, but currently it's not possible(

**Comment by APshenkin** [comment]
@pimterry is there any blockers to merge this feature and update mockttp?

**Comment by pimterry** [comment]
No specific blockers, I just need time to sit down and review it again. Life's got a bit hectic, but I should be able to sort this this weekend.

**Comment by pimterry** [comment]
I've made a couple more tweaks here, but otherwise this all looks good to me now - I'll merge in a sec once the tests pass.

**Pull request merged**